File: max_subarray.c
// find out the sum of the "maximum subarray"

#include <stdio.h>

int max_subarray (int arr[], int size)
{
	int max;
	int tmp_sum = 0;
	int i;

	if (0 >= size) return 0xffffffff;
	if (1 == size) return arr[0];

	max = arr[0];
	for (i = 0; i < size; i++) {
		if (tmp_sum < 0) tmp_sum = arr[i]; // reset start point
		else tmp_sum += arr[i];            // try this one makes sum still >0 or not
		if (max < tmp_sum) max = tmp_sum;
	}
	
	return max;
}
int main()
{
	int arr[] = {1,-2,-4,-4,100};
	int sum = max_subarray(arr, sizeof(arr) / sizeof(int));
	printf("sum = %d\n", sum);
	
	return 0;
}

File: maxSubArray.java
import java.lang.Integer;

public class Solution {
    public int maxSubArray(int[] A) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if (null == A || A.length == 0)
            return 0;
        
        int sum = -1;
        int max = Integer.MIN_VALUE;
        
        for (int i = 0; i < A.length; i++) {
            sum = sum >= 0 ? sum + A[i] : A[i];
            max = sum > max ? sum : max;
        }
        
        return max;
    }
}
